File: Data Structures and Algorithms/AI/Classification/Decision Tree/Gini Index/Cpp/gini.cpp
#include <vector>
#include <set>
#include <limits>
#include <type_traits>
#include <algorithm>

/**
 * @brief Alias for the dataset type. Each row is a vector of values, with the last value being the class label.
 * @tparam T Numeric data type (e.g., int, float, double)
 */
template<typename T>
using dataset_t = std::vector<std::vector<T>>;

/**
 * @brief Computes the Gini index for a split dataset.
 * @tparam T Numeric data type
 * @param groups Pair of split datasets (left and right)
 * @param classes Set of unique class labels
 * @return Gini index value indicating impurity (lower is better)
 */
template<typename T>
T gini_index(const std::pair<dataset_t<T>, dataset_t<T>>& groups, const std::set<T>& classes) {
    T total_size = static_cast<T>(groups.first.size() + groups.second.size());
    T gini = 0;

    for (const auto& group : {groups.first, groups.second}) {
        size_t size = group.size();
        if (size == 0) continue;

        T score = 0;
        for (T class_val : classes) {
            size_t count = std::count_if(group.begin(), group.end(),
                [class_val](const std::vector<T>& row) {
                    return row.back() == class_val;
                });
            T p = static_cast<T>(count) / static_cast<T>(size);
            score += p * p;
        }

        gini += (1 - score) * (static_cast<T>(size) / total_size);
    }

    return gini;
}

File: Data Structures and Algorithms/AI/Classification/Decision Tree/Gini Index/Python/gini.py
def gini_index(groups, classes):
    """
    @brief Calculate the Gini index for a split dataset.

    @param groups A list of groups (each group is a list of rows) resulting from a dataset split.
    @param classes A list of all unique class labels in the dataset.

    @return The Gini index as a float value. Lower Gini indicates better purity.
    """
    n_instances = float(sum([len(group) for group in groups]))
    gini = 0.0
    for group in groups:
        size = len(group)
        if size == 0:
            continue

        score = 0.0
        for class_val in classes:
            # Proportion of class_val in the group
            p = [row[-1] for row in group].count(class_val) / size
            score += p * p
        # Weight the group’s Gini score by its relative size
        gini += (1.0 - score) * (size / n_instances)

    return gini
